### PR TITLE
Permissions API and WebBluetooth integration

### PR DESCRIPTION
**Pull request by dati91**
This implements the [Permissions API](https://w3c.github.io/permissions/) spec.
Also includes the WebBluetooth related implementation for this.

There are some know issues:
- [ ] If the descriptor name is invalid [this](https://gist.github.com/dati91/7a6a0a563d90f49ba5a351e48c5b626b#file-permissionstatusbindings-rs-L323) will throw an error, rather that return it and we could handle it.
- [x] The [environment settings object](https://html.spec.whatwg.org/multipage/webappapis.html#environment-settings-object) is not implemented in servo and the spec rely on it.
- [x] There is a popup in the implementation which prevent us to add wpt test, we should figure out a way to make it work
- [ ] The allowedDevice's allowed_services attribute is not used in our implementation, because we store these in the lower level, not in the dom side.
- [ ] We think the bluetooth revoke function will need some more work, but the problem is the spec needs clarifications on that part.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because 

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15314)

### CODE REVIEW

File: components/script/dom/bluetooth.rs [hunk 1]
```diff
@@ -481,3 +549,148 @@ impl AsyncBluetoothListener for Bluetooth {
...
+        }
+    }
+
+    #[allow(unrooted_must_root)]
+    // https://webbluetoothcg.github.io/web-bluetooth/#query-the-bluetooth-permission
+    fn permission_query(cx: *mut JSContext, promise: &Rc<Promise>,
+                        descriptor: &BluetoothPermissionDescriptor,
+                        status: &BluetoothPermissionResult) {
+        // Step 1.
+        // TODO: `environment settings object` is not implemented in Servo yet.
```
Comment (Ms2ger) [inline components/script/dom/bluetooth.rs]: The "relevant global object" for *status* is just `status.global()` in Servo.

Comment (dati91) [inline components/script/dom/bluetooth.rs]: Great! I based this comment on [this](https://github.com/servo/servo/blob/master/components/script/dom/request.rs#L139) line. Maybe next time i shouldn't give up on searching that early. :)

File: components/script/dom/permissions.rs [hunk 2]
```diff
@@ -276,6 +266,23 @@ pub fn get_descriptor_permission_state(permission_name: PermissionName ,
...
+                                              YesNo::No) {
+        YesNo::Yes => return PermissionState::Granted,
+        YesNo::No => return PermissionState::Denied,
+    };
+}
+
+#[cfg(not(target_os = "linux"))]
+fn prompt_user(_permission_name: PermissionName) -> PermissionState {
+    // TODO popup only supported on linux
+    PermissionState::Denied;
```
Comment (zakorgy) [inline components/script/dom/permissions.rs]: Remove the `;` from the end of line. It causes the build error on windows: https://ci.appveyor.com/project/servo/servo/build/1.0.13207/job/f0pbavepdah2u611#L619

File: components/script/dom/permissions.rs [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                          promise: &Rc<Promise>,
+                          descriptor: &PermissionDescriptor,
+                          status: &PermissionStatus) {
+        // Step 1.
+        Permissions::permission_query(cx, promise, descriptor, status);
+
+        // TODO: Step 2 - 4: `environment settings object` is not implemented in Servo yet.
+        // For this reason in the `get_descriptor_permission_state` function we can't decide
+        // if we have a secure context or not, or store the previous invocation results.
+        // Without these the remaining steps can't be implemented properly.
```
Comment (jdm) [inline components/script/dom/permissions.rs]: We can use `GlobalScope::current()` to give us the equivalent of the `current environment settings object`.

File: components/script/dom/permissions.rs [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    fn permission_revoke(_descriptor: &PermissionDescriptor, _status: &PermissionStatus) {}
+}
+
+// https://w3c.github.io/permissions/#permission-state
+pub fn get_descriptor_permission_state(permission_name: PermissionName ,
```
Comment (jdm) [inline components/script/dom/permissions.rs]: No need for this to be public.

File: components/script/dom/permissions.rs [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+                                       _env_settings_obj: Option<*mut JSObject>)
```
Comment (jdm) [inline components/script/dom/permissions.rs]: This should be `Option<&GlobalScope>` instead.

File: components/script/dom/permissions.rs [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+                                       -> PermissionState {
+    // TODO: Step 1: If settings wasn’t passed, set it to the current settings object.
+    // TODO: `environment settings object` is not implemented in Servo yet.
```
Comment (jdm) [inline components/script/dom/permissions.rs]: `GlobalScope::current()`.

File: components/script/dom/permissions.rs [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+                YesNo::Yes => return PermissionState::Granted,
+                YesNo::No => return PermissionState::Denied,
+            };
+        } else {
+            return PermissionState::Denied;
+        }
+    }
+
+    // TODO: Step 3: Store the invocation results
+    // TODO: `environment settings object` is not implemented in Servo yet.
```
Comment (jdm) [inline components/script/dom/permissions.rs]: We can store this in a field on Window.

**Review by jdm — CHANGES_REQUESTED** [review]
Sorry for the delay! I've been really busy and there was a _lot_ of code to review.

File: components/script/dom/webidls/BluetoothPermissionResult.webidl [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+};
+
+dictionary AllowedBluetoothDevice {
+  required DOMString deviceId;
+  required boolean mayUseGATT;
+  // An allowedServices of "all" means all services are allowed.
+  required (DOMString or sequence<UUID>) allowedServices;
+};
+
+dictionary BluetoothPermissionData {
```
Comment (jdm) [inline components/script/dom/webidls/BluetoothPermissionResult.webidl]: Given that BluetoothPermissionData and AllowedBluetoothDevice are never exposed to web content as far as I can see, we don't need to define them in WebIDL. They can be native Rust types instead, which will make it easier to derive implementations like HeapSizeOf and JSTraceable on them.

File: components/script/dom/bluetooth.rs [hunk 9]
```diff
@@ -100,6 +79,46 @@ impl Clone for AllowedBluetoothDevice {
     }
 }
 
+#[derive(HeapSizeOf, JSTraceable)]
+pub struct BluetoothExtraPermissionData {
+    permission_data: DOMRefCell<BluetoothPermissionData>,
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: We should inline the members of BluetoothPermissionData inside BluetoothExtreaPermissionData; there's no need for the extra level of indirection.

File: components/script/dom/bluetooth.rs [hunk 10]
```diff
@@ -100,6 +79,46 @@ impl Clone for AllowedBluetoothDevice {
...
+            permission_data: DOMRefCell::new(BluetoothPermissionData { allowedDevices: Vec::new() }),
+        }
+    }
+
+    pub fn add_new_allowed_device(&self, allowed_device: AllowedBluetoothDevice) {
+        self.permission_data.borrow_mut().allowedDevices.push(allowed_device);
+    }
+
+    fn get_allowed_devices(&self) -> Vec<AllowedBluetoothDevice> {
+        self.permission_data.borrow().allowedDevices.clone()
```
Comment (jdm) [inline components/script/dom/bluetooth.rs]: If we return `Ref<Vec<...>>` instead (via Ref::map) then we don't need to clone the array every time.

**Review by jdm — CHANGES_REQUESTED** [review]

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
From my perspective these changes look good. Was there anything else you intended to change before merging?

**Comment by dati91** [comment]
No, that was all.

**Pull request merged**